### ISSUE TITLE
Autobahn-Java on Android Connection Timeout

### ISSUE THREAD

**Issue by gammazero** [gammazero/nexus#146]
Originally posted by @jurjen96 in #143:

We are currently trying to implement a wamp connection between Autobahn-Java and the Nexus router. When we try to connect to the router, we get, after 5 seconds, the error that a timeout occurred:

`2018/08/08 12:19:10 Error attaching to router: did not receive HELLO: timeout waiting for message`

A code snippet for connecting to the router:
```
private void onJoinCallback(Session wamp_session, SessionDetails details){
        System.out.println("Successfully connected to Realm");
}

public void connect(){
        session.addOnJoinListener(this::onJoinCallback);
        client = new Client(session, address, realm);
        CompletableFuture<ExitInfo> exitInfoCompletableFuture = client.connect();
}
```

We also tried to connect to the router of Crossbar, and we got a successful connection between the session and realm. Do you know if Autobahn-Java is compatible with the Nexus router?

when we ran our code, we didn't use the example code of Autobahn-Java. By the way, we are running the Java code in Android.

> Also, if you enable debug in the router, does it log anything at all? This would indicate if the client was at least able to connect.

When we enable debug in the router, we see with our Android Java code only the Hello timeout waiting for message error. When we use the example code in a Java application, we get the following response from the Nexus router:
`2018/08/09 11:59:36 Error attaching to router: did not receive HELLO: receive channel closed`
And on the client side:
`Aug 09, 2018 11:59:33 AM io.crossbar.autobahn.wamp.transports.NettyWebSocketClientHandler
INFO: WebSocket Client disconnected!
Aug 09, 2018 11:59:33 AM ExampleClient onDisconnectCallback
INFO: Session with ID=0, disconnected.`

For the example of Autobahn Java we didn't use the docker version, but we copied it to a Java application because we had some difficulties running their docker version. For the example of Autobahn Java we configured the router to use port 8080, but we still get "receive channel closed". 

> What transport are you using Websocket, RawSocket (TCP, Unix)?

For both the example and Android version we use Websocket transports

> What interface:port is server listening on, client connecting to (ipv6-v4 used)?

We use our local network address 192.168.0.x and port 8080. Our client is also on our local network.

> Using TLS?

No

> Are you setting WebsocketServer.KeepAlive to a non-zero value?

We set the keepalive to 5 seconds to detect if one of our clients disconnects. When we set the keepalive parameter to 0, the error of "timeout waiting for message" still persists.

Output we got from the client, sorry about that. Might be helpfull:

```
W/art: Method processed more than once: boolean io.crossbar.autobahn.websocket.WebSocketReader.processHandshake()
W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.hashCode()' on a null object reference
                  at io.crossbar.autobahn.wamp.transports.AndroidWebSocket.initializeSerializer(AndroidWebSocket.java:160)
W/System.err:     at io.crossbar.autobahn.wamp.transports.AndroidWebSocket.access$100(AndroidWebSocket.java:32)
                  at io.crossbar.autobahn.wamp.transports.AndroidWebSocket$1.onConnect(AndroidWebSocket.java:86)
                  at io.crossbar.autobahn.websocket.WebSocketConnection$2.handleMessage(WebSocketConnection.java:648)
W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
                  at android.os.Looper.loop(Looper.java:148)
                  at android.app.ActivityThread.main(ActivityThread.java:5509)
W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:772)
                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:662)
```

**Comment by om26er** [comment]
@jurjen96, I just stumbled up this issue. Could you clarify if your app is supposed to run on Android or a server ? I am asking because we, on Android use our own websocket implementation and for servers we use Netty.

The autobahn-java implementation of WAMP was mostly based/inspired by autobahn-python' implementation and I only tested it with Crossbar, I will check later if I can reproduce your issue, could you provide a code snippet that I could use ?

**Comment by jurjen96** [comment]
Our app is supposed to run on Android.

Currently, I don't have a code snippet at hand since we moved to autobahn-js in an android app. If necessary, I can have a look at a code snippet next week.

**Comment by jurjen96** [comment]
I went back to our old commits and I was unable to reproduce the issue described above. I tried to use 
`autobahn-android-legacy:18.5.1` and `jawampa-core:0.5.0` with `jawampa-netty:0.5.0` and both approaches did not result in a timeout. This ticket can be closed. Thank you guys for your support!

**Comment by gammazero** [comment]
Closing issue based on last comment by @jurjen96

**Comment by robfra31** [comment]
Hello, I would like to know how the issue was resolved @jurjen96  ?
For me, the `Error attaching to router: did not receive HELLO: timeout waiting for message` appears only when I set the `"enable_compression": true,` in the Nexus JSON configuration.

@gammazero: What is exactly the purpose of this attibute `enable_compression`, is it for enabling MsgPack in Nexus router ? Does it have a lot of impact of the performances on the transmitting time of the data ?

**Comment by martin31821** [comment]
enable_compression should enable permessage-deflate, i.e. gzip per message. msgpack is handled at the protocol level by setting allowed subprotocols to `wamp.2.msgpack`.

**Comment by robfra31** [comment]
Okay tank you for your answer. 
Do you have any idea why this `enable_compression` create an error when the Client (Autobahn-java) is connecting to the server ?

**Comment by jurjen96** [comment]
@robfra31 I'm not 100% sure how we resolved this issue since we moved away from using Autobahn Java and Jawampa to Autobahn|JS. My best guess would be that we updated/changed some settings in the router which fixed our issue. It might be that we, as you suggested, changed the ```enable_compression``` while we were converting our code from Java to JS, but I can't say for sure since we changed a lot of things on the JavaScript side and the Nexus router side.

**Comment by robfra31** [comment]
@jurjen96 Thank your answer, I will try from my part.

@martin31821 I would like to know how the use of MsgPack via wamp.2.msgpack is defined in the Nexus router configuration ? Is it a special parameter to set in the JSON file ?

**Comment by martin31821** [comment]
@robfra31 if your client announces msgpack, the router should pick it up implicitly. (I'm not sure for nexusd, but our nexus-embedding router does this).

**Comment by robfra31** [comment]
@martin31821 Do you know the strategy for choosing the serializer in Nexus ? Is it by order in the list of the possible serializers (the first is chosen if possible, if not it is the second, etc..) or is it according to the other client already connected to the realm ?
For example: 
- A client is connecting to the Nexus router with a MsgPack serialization
- Another client joins the realm and announces 2 serializers (JSON and MsgPack), does the nexus server will pick MsgPack too because there is already another client in MsgPack or does it choose JSON because it is the first declare ?

And do you know if 2 clients are "speaking" in MsgPack, does the server will make an internal decompression and re-compression between the receive of the message from the publisher and the transmit of the message to the subscriber ?

Thank you very much !

**Comment by martin31821** [comment]
For the first question: should be the first serializer which both parties support.
Second question: Yes, the router will always deserialize all messages, process them and then reencode them.

**Comment by jurjen96** [comment]
Hey guys, for another project we have to use Autobahn-Java and we encountered the same problem again as described above (`did not receive HELLO`). After some digging through the Autobahn Java code, we found out that there is an issue with negotiation which serializer to use. On the [onConnect](https://github.com/crossbario/autobahn-java/blob/master/autobahn/src/main/java/io/crossbar/autobahn/wamp/transports/AndroidWebSocket.java#L86) of Autobahn-Java, it tries to read the protocol the client should use. However, the response.protocol is not set and returns a null which causes the issue described above (`W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.hashCode()' on a null object reference`). Is this an issue with the implementation of the Nexus router or Autobahn-Java?
